typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1015 @@
+
+### Strictness
+
+Unlike Haskell, TypeScript does not have a focus on types.
+They are just there to help programmers write stricter and more understandable code, making their life easier.
+You can use types like `any`, and nobody pushes you into strict frameworks.
+
+Nevertheless, with a bunch of compiler options, you can make TypeScript more strict in types and other things.
+To do that, you need to add the `strict` compiler option to your `tsconfig.json` file.
+It will enable all strictness-related flags for you, and if in the future something will be added to the strict options list, you will not need to add it manually.
+But in case you want to specify them one-by-one, here are the current ones: alwaysStrict, strictNullChecks, strictBindCallApply, strictFunctionTypes, strictPropertyInitialization, noImplicitAny, noImplicitThis, useUnknownInCatchVariables.
+
+### Type aliases and newtypes
+
+#### Type aliases
+
+Type aliases work the same as in Haskell – you can create a new name to refer to some type.
+
+```typescript
+type Email = string;
+```
+
+And you can also create types with properties.
+
+```typescript
+type Document = {
+  name: string;
+  athor?: string;
+}
+```
+
+Symbol `?` here is used as syntactic sugar for the `string | undefined` type.
+
+#### Newtype
+
+Creating a `newtype` is not so trivial.
+There is no such analogue in TypeScript, but you can do it by using tags.
+
+```typescript
+type Email = string & { readonly __tag: unique symbol };
+type City = string & { readonly __tag: unique symbol };
+```
+
+Here's how all this magic works:
+
+* With [intersection types](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#intersection-types) (and syntax), you can add a `tag` field to your type.
+* `symbol` lets you declare const-named properties on types.
+* `unique` means that the type is unique – your type is tagged with a unique symbol.
+
+By using [`as`](https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions) (which does type assertion), you can cast your base type to a tagged type.
+And when you try to pass a value of another tagged type, the compiler says that the two unique symbols are different.
+
+```typescript
+function sendMessage(email: Email);
+
+sendMessage("message"); // Argument of type 'string' is not assignable to parameter of type 'Email'.
+sendMessage("St. Petersburg" as City); // Argument of type 'City' is not assignable to parameter of type 'Email'.
+
+sendMessage("email@gmail.com" as Email); // Ok
+```
+
+Anyway, this is not an ideal solution for newtypes.
+There is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution.
+And there is also a [library](https://github.com/sveseme/taghiro), which provides a prettier implementation for them.
+
+### Unit types, unions, data types, and pattern matching
```

Comment:
Could this section be called ADTs in TypeScript?

Reply:
Not sure if it will be a right definition but in article title context I think we can do it

Reply:
Maybe just ADTs or Algebraic data types? To not explicitly mention TypeScript

Reply:
Yeah, that is fine.